File: python/guessthenumber/evaluator.py
from dataclasses import dataclass


@dataclass
class Answer:

    number: str
    rights: int
    wrongs: int
    present_but_wrong: int

    def __str__(self) -> str:
        return f'Rights: {self.rights}. Wrongs: {self.wrongs}. Present but wrong: {self.present_but_wrong}.'

    def __bool__(self) -> bool:
        return len(self.number) == self.rights and self.wrongs == 0 and self.present_but_wrong == 0


class Evaluator:

    @classmethod
    def evaluate_guess(cls, guess: str, number: str) -> Answer:
        rights: int = 0
        wrongs: int = 0
        present_but_wrong: int = 0

        if guess == number:
            return Answer(number, len(number), 0, 0)

        for idx, char in enumerate(guess):
            if char == number[idx]:
                rights += 1
            elif char in number:
                present_but_wrong += 1
            else:
                wrongs += 1

        return Answer(number, rights, wrongs, present_but_wrong)

File: rust/guessthenumber/src/game/evaluator.rs
extern crate rand;


pub struct Answer {
    rights: u8,
    wrongs: u8,
    present_but_wrong: u8
}

impl Answer {
    pub fn is_correct(&self, guess: String) -> bool {
        self.rights == guess.len() as u8
        && self.wrongs == 0
        && self.present_but_wrong == 0
    }
}

pub fn evaluate_guess(guess: &str, number: String) -> Answer {
    let mut rights: u8 = 0;
    let mut wrongs: u8 = 0;
    let mut present_but_wrong: u8 = 0;

    let mut iterable_number = number.chars();

    for c in guess.chars()
    {
        let is_right: bool = c == iterable_number.next().unwrap();
        let is_present: bool = number.contains(c);

        match (is_right, is_present) {
            (false, false) => wrongs += 1,
            (false, true)  => present_but_wrong += 1,
            (true, _)      => rights += 1,
        }
    };

    Answer { rights, wrongs, present_but_wrong }
}
